### PR TITLE
Added `non-printable-ascii` lint

### LINKED ISSUE #63682: Warn about non-printable characters in string literals
Currently a string literal with control characters like `\0` or `\v` is accepted without any warnings. The only exception is `\r`, which gives a hard error.

It makes more sense to treat all non-[printable](https://en.wikipedia.org/wiki/ASCII#Printable_characters), non \t, \n ASCII characters as a warning. 

Steps to fix:

1. Add  `NonPrintableAsii` to [EscapeError](https://github.com/rust-lang/rust/blob/ef1ecbefb8719e408150738664d443a73f757ffd/src/librustc_lexer/src/unescape.rs#L11)
2. Produce this error somewhere around [here](https://github.com/rust-lang/rust/blob/ef1ecbefb8719e408150738664d443a73f757ffd/src/librustc_lexer/src/unescape.rs#L135-L137)
3. Add lexer-level [tests](https://github.com/rust-lang/rust/blob/ef1ecbefb8719e408150738664d443a73f757ffd/src/librustc_lexer/src/unescape/tests.rs)
4. Handle this "error" in [`unespcape_error_reporting`](https://github.com/rust-lang/rust/blob/ef1ecbefb8719e408150738664d443a73f757ffd/src/libsyntax/parse/unescape_error_reporting.rs#L38). Note that, unlike other real errors, this one should be just a warning.
5. Adjust the affected ui tests

I am not sure how to make this warning work with `#[allow]` lint infrastructure: we definitely can't do this in the lexer.

### PR DESCRIPTION
**Pull request by horse-princess**
Closes #63682.

### CODE REVIEW

**Comment by Centril** [comment]
Perf is clean :)

**Comment by petrochenkov** [comment]
Is `is_ascii_control` the right thing to use here?
Perhaps some unicode category would be more suitable here?
(This affects the lint name as well.)

**Comment by Areredify** [comment]
сс @matklad

**Comment by JohnCSimon** [comment]
Ping from triage:
@Manishearth Can you please review this PR?
Thank you.

File: src/librustc_parse/lexer/unescape_error_reporting.rs [hunk 1]
```diff
@@ -201,6 +202,40 @@ pub(crate) fn emit_unescape_error(
...
+    sess: &ParseSess,
+    // full span of the literal, including quotes
+    span_with_quotes: Span,
+    // interior part of the literal, without quotes
+    character: char,
+    range: Range<usize>,
+) {
+    const ALLOWED_ASCII_CONTROL_CHARS: [char; 2] = ['\t', '\n'];
+
+    if range.len() != 1 {
```
Comment (Manishearth) [inline src/librustc_parse/lexer/unescape_error_reporting.rs]: Why this check? Please add a comment.

File: src/librustc_parse/lexer/unescape_error_reporting.rs [hunk 2]
```diff
@@ -201,6 +202,40 @@ pub(crate) fn emit_unescape_error(
...
+        return;
+    }
+    if !ALLOWED_ASCII_CONTROL_CHARS.contains(&character) && character.is_ascii_control() {
+        sess.buffer_lint(
+            BufferedEarlyLintId::NonPrintableAscii,
+            span_with_quotes,
+            ast::CRATE_NODE_ID,
+            "non-printable ASCII character in literal"
+        )
```
Comment (Manishearth) [inline src/librustc_parse/lexer/unescape_error_reporting.rs]: We should suggest they use an escape code.

**Review by Manishearth — APPROVED** [review]
The tests should ensure that strings with escape codes work.

**Comment by Areredify** [comment]
@JohnCSimon Sorry for the late response. Original issue talks specifically about ASCII characters, IMO it makes sense to extend it to `Cc` characters, but I would like to hear @matklad opinion on this.

**Comment by joelpalmer** [comment]
Ping from Triage:  any updates @Areredify @matklad?

**Comment by joelpalmer** [comment]
Ping from Triage: any updates @Areredify @matklad?

**Comment by JohnCSimon** [comment]
Triage: 
@Areredify  - unfortunately this PR hasn't seen any movement for over a month, so I'm closing this as inactive. 
Thanks for your efforts, and feel free to reopen this PR when you have the time to update this and fix the merge conflicts.

Please don't push to the PR while it is closed as that prevents it from being reopened.

**Pull request closed without merging**